src/gurt/telemetry.c

```diff
@@ -0,0 +1,2341 @@
+		p = path;
+		rc = d_tm_build_path(&p, item, args);
+		va_end(args);
+
+		if (rc != D_TM_SUCCESS)
+			goto failure;
+		node = d_tm_find_metric(d_tm_shmem_root, path);
+		if (metric != NULL)
+			*metric = node;
+	}
+
+	if (node == NULL) {
+		rc = d_tm_add_metric(&node, path, D_TM_COUNTER, "N/A", "N/A");
+		if (rc != D_TM_SUCCESS) {
+			D_ERROR("Failed to add and incremement counter [%s]: "
+				"rc = %d\n", path, rc);
+			goto failure;
+		}
+		if (metric != NULL)
+			*metric = node;
+	}
+
+	if (node->dtn_type == D_TM_COUNTER) {
+		D_MUTEX_LOCK(&node->dtn_lock);
+		node->dtn_metric->dtm_data.value++;
+		D_MUTEX_UNLOCK(&node->dtn_lock);
+	} else {
+		rc = -DER_OP_NOT_PERMITTED;
+		D_ERROR("Failed to increment counter [%s] on item not a "
+			"counter.  Operation mismatch: rc = %d\n",
+			node->dtn_name, rc);
+		goto failure;
+	}
+	return rc;
+
+failure:
+	return rc;
+}
+
+
+/**
+ * Record the current timestamp
+ *
+ * The timestamp is specified either by an initialized pointer or by a fully
+ * qualified item name.  If an initialized pointer is provided, the metric is
+ * accessed directly.  Otherwise, a lookup is performed on the path name
+ * provided in order to find the specified item.  If the item cannot be found,
+ * it is created.  If the item is created, the callers pointer is initialized
+ * for this item.  The pointer is used for direct access on subsequent calls for
+ * faster access.
+ *
+ * \param[in,out]	metric	Pointer to the metric
+ * \param[in]		item	Full path name to this \a item
+ *				If supplied, the path name is specified by the
+ *				\a item and any strings following it.  The last
+ *				item must be NULL.
+ *
+ * \return			D_TM_SUCCESS		Success
+ *				-DER_EXCEEDS_PATH_LEN	The full name length is
+ *							too long
+ *				-DER_OP_NOT_PERMITTED	Operation not permitted
+ *							because the \a item is
+ *							not a timestamp
+ *				-DER_UNINIT		API not initialized
+ *				-DER_INVAL		\a metric and \a item
+ *							are NULL
+ */
+int
+d_tm_record_timestamp(struct d_tm_node_t **metric, char *item, ...)
+{
+	struct d_tm_node_t	*node = NULL;
+	char			path[D_TM_MAX_NAME_LEN] = {};
+	int			rc = D_TM_SUCCESS;
+
+	if (d_tm_shmem_root == NULL)
+		return -DER_UNINIT;
+
+	if ((metric != NULL) && (*metric != NULL)) {
+		node = *metric;
+	} else {
+		va_list	args;
+		char	*p;
+
+		if (item == NULL) {
+			rc = -DER_INVAL;
+			goto failure;
+		}
+
+		va_start(args, item);
+		p = path;
+		rc = d_tm_build_path(&p, item, args);
+		va_end(args);
+		if (rc != D_TM_SUCCESS)
+			goto failure;
+		node = d_tm_find_metric(d_tm_shmem_root, path);
+		if (metric != NULL)
+			*metric = node;
+	}
+
+	if (node == NULL) {
+		rc = d_tm_add_metric(&node, path, D_TM_TIMESTAMP, "N/A", "N/A");
+		if (rc != D_TM_SUCCESS) {
+			D_ERROR("Failed to add and record timestamp [%s]: "
+				"rc = %d\n", path, rc);
+			goto failure;
+		}
+		if (metric != NULL)
+			*metric = node;
+	}
+
+	if (node->dtn_type == D_TM_TIMESTAMP) {
+		D_MUTEX_LOCK(&node->dtn_lock);
```

Comment:
We are talking about removing all of the locks that protect the metrics.  If that happens, this one will go at the same time.  Otherwise, I'll keep it as-is until then.